src/preset.ts

```diff
@@ -0,0 +1,1222 @@
+
+/**
+ * Call configurePresets, buildPresets, or testPresets to get the latest presets when thie event is fired.
+ */
+export function onUserPresetsChanged(listener: () => any) { return userPresetsChangedEmitter.event(listener); }
+
+export function getPresetByName<T extends Preset>(presets: T[], name: string): T | null {
+  for (const preset of presets) {
+    if (preset.name === name) {
+      return preset;
+    }
+  }
+  return null;
+}
+
+function isInheritable(key: keyof ConfigurePreset | keyof BuildPreset | keyof TestPreset) {
+  return key !== 'name' && key !== 'hidden' && key !== 'inherits' && key !== 'description' && key !== 'displayName';
+}
+
+function removeNullEnvVars(preset: Preset) {
+  if (preset.environment) {
+    for (const key in preset.environment) {
+      if (preset.environment[key] === null) {
+        delete preset.environment[key];
+      }
+    }
+  }
+}
+
+let kits: Kit[] = [];
+
+/**
+ * Using kits as compilers
+ */
+export function setCompilers(_kits: Kit[]) {
+  kits = _kits;
+}
+
+/**
+ * Used for both expandConfigurePreset and expandVendorForConfigurePreset
+ */
+const referencedConfigurePresets: Set<string> = new Set();
+
+/**
+ * This is actually a very limited version of expandConfigurePreset.
+ * Build/test presets currently don't need this, but We could extend this
+ * to work with build/test presets in the future.
+ * Use expandVendorPreset if other fields are needed.
+ * They should NOT be used together.
+ * They should Not call each other.
+ */
+export function expandVendorForConfigurePresets(): void {
+  for (const preset of configurePresets()) {
+    getVendorForConfigurePreset(preset.name);
+  }
+  for (const preset of userConfigurePresets()) {
+    getVendorForConfigurePreset(preset.name);
+  }
+}
+
+function getVendorForConfigurePreset(name: string): VendorType | Vendor_VsSettings | null {
+  referencedConfigurePresets.clear();
+  return getVendorForConfigurePresetImpl(name);
+}
+
+function getVendorForConfigurePresetImpl(name: string, allowUserPreset: boolean = false): VendorType | Vendor_VsSettings | null {
+  let preset = getPresetByName(configurePresets(), name);
+  if (preset) {
+    return getVendorForConfigurePresetHelper(preset);
+  }
+
+  if (allowUserPreset) {
+    preset = getPresetByName(userConfigurePresets(), name);
+    if (preset) {
+      return getVendorForConfigurePresetHelper(preset, true);
+    }
+  }
+
+  return null;
+}
+
+function getVendorForConfigurePresetHelper(preset: ConfigurePreset, allowUserPreset: boolean = false): VendorType | Vendor_VsSettings | null {
+  if (preset.__expanded) {
+    return preset.vendor || null;
+  }
+
+  if (referencedConfigurePresets.has(preset.name)) {
+    // Refernced this preset before, but it doesn't have a configure preset. This is a circular inheritance.
+    log.error('circular.inherits.in.config.preset', 'Circular inherits in configure preset {0}', preset.name);
+    return null;
+  }
```

Comment:
"referenced"... see it in 5 more places throughout this PR, no need to point them individually.